pprint/src-0/TPrintImpl.scala

```diff
@@ -0,0 +1,104 @@
+      '{ $cfg.typeColor($s).render }
+    }
+
+    def printSymString(cfg: Expr[TPrintColors], s: String) =
+      if (s.toString.startsWith("_$")) "_"
+      else s.toString.stripSuffix(".type")
+
+    def printBounds(cfg: Expr[TPrintColors])(lo: Type, hi: Type) = {
+      val loTree =
+        if (lo =:= typeOf[Nothing]) None else Some('{ " >: " + ${ rec0(cfg)(lo) }})
+      val hiTree =
+        if (hi =:= typeOf[Any]) None else Some('{ " <: " + ${ rec0(cfg)(hi) }})
+      loTree.orElse(hiTree).map((expr) => '{ "_" + $expr }).getOrElse('{ "_" })
+    }
+
+    def printSym(cfg: Expr[TPrintColors], s: String): Expr[String] = {
+      val expr = Expr(s)
+      literalColor(cfg, '{ fansi.Str($expr) })
+    }
+
+    //TODO: We don't currently use this method
+    def prefixFor(cfg: Expr[TPrintColors])(pre: TypeTree, sym: String): Expr[String] = {
+      // Depending on what the prefix is, you may use `#`, `.`
+      // or even need to wrap the prefix in parentheses
+      val sep = pre match {
+        case x if x.toString.endsWith(".type") =>
+          '{ ${rec0(cfg)(pre.tpe)} + "." }
+      }
+      '{ $sep + ${ printSym(cfg, sym) } }
+    }
+
+
+    def printArgs(cfg: Expr[TPrintColors])(args: List[TypeOrBounds]): Expr[String] = {
+      val added = args.map {
+        case tpe: Type =>
+          rec0(cfg)(tpe, false)
+        case TypeBounds(lo, hi) =>
+          printBounds(cfg)(lo, hi)
+      }.reduceLeft { (l, r) =>
+        '{ $l  + ", " + $r }
+      }
+      '{ "[" + $ { added } + "]" }
+    }
+
+
+    object RefinedType {
+      def unapply(tpe: Type): Option[(Type, List[(String, TypeOrBounds)])] = tpe match {
+        case Refinement(p, i, b) =>
+          unapply(p).map {
+            case (pp, bs) => (pp, (i -> b) :: bs)
+          }.orElse(Some((p, (i -> b) :: Nil)))
+        case _ => None
+      }
+    }
+
+    def rec0(cfg: Expr[TPrintColors])(tpe: Type, end: Boolean = false): Expr[String] = tpe match {
+      case TypeRef(NoPrefix(), sym) =>
+        printSym(cfg, sym)
+        // TODO: Add prefix handling back in once it works!
+      case TypeRef(_, sym) =>
+        printSym(cfg, sym)
+      case AppliedType(tpe, args) =>
+        '{ ${ printSym(cfg, tpe.typeSymbol.name) } + ${ printArgs(cfg)(args) } }
+      case RefinedType(tpe, refinements) =>
+        val pre = rec0(cfg)(tpe)
+        lazy val defs = refinements.collect {
+          case (name, tpe: Type) =>
+            '{ "type " + ${ Expr(name) } + " = " + ${ rec0(cfg)(tpe) } }
+          case (name, tpe: Type) =>
+            '{ "type " + ${ Expr(name) } + " = " + ${ rec0(cfg)(tpe) } }
+        }.reduceLeft { (l, r) =>
+          '{ $l + "; " + $r }
+        }
+        '{ $pre + ${ if(refinements.isEmpty) '{ "" } else '{ "{" + $defs + "}" } } }
+      case _ =>
+        Expr(t.show)
+    }
+    '{
+      new TPrint[T] {
+        final def render(implicit cfg: TPrintColors): String = ${ rec0('cfg)(t.unseal.tpe) } 
```

Comment:
Not sure how many string additions `req0` might generate, but this will have a quadratic behavior with might quickly become slow. Maybe we should have a `StringBuilder` and instead of string additions, we would append to the builder.

```scala
        final def render(implicit cfg: TPrintColors): String = 
          val sb: StringBuffer = ...
          ${ rec0('cfg, 'sb)(t.unseal.tpe) } 
          sb.result()
```

Reply:
Perhaps I'm misunderstanding, but since the code generates an inlined expression like `a + b + c`, Java should automatically switch it to a `StringBuilder` no?

Reply:
At any rate; I'd like to keep this implementation as similar to the scala 2 code as possible so it's easy to port the behaviour between them.

Reply:
You might be right, we probably optimize it. Would be good to double check to see if the optimization is done properly.